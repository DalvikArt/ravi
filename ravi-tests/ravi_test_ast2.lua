local function doast(str)
    local x, msg = ast.parse(str)
    if not x then error(msg) end
    print(x:tostring())
end

local str =
[[return function()
    local a = 1

	local function y()
		return function()
			return a
		end
	end

	local a = 5

	local function z()
		return function()
			return a
		end
	end

	return y, z
end
]]
doast(str)

str =
[[return function (a, b, c, d, e)
    if a == b then goto l1
    elseif a == c then goto l2
    elseif a == d then goto l2
    else if a == e then goto l3
        else goto l3
        end
    end
    ::l1:: ::l2:: ::l3:: ::l4::
end
]]
doast(str)

str =
[[return function (a) while a < 10 do a = a + 1 end end
]]
doast(str)

-- Note that the while loop above is supposed to ultimately
-- generate same code as below
str=
[[return function (a) ::L2:: if not(a < 10) then goto L1 end; a = a + 1;
                  goto L2; ::L1:: end
]]
doast(str)

str=
[[return function ()
    (function () end){f()}
  end
]]
doast(str)

str=
[[return function ()
    local sum 
    for j = 1,500 do
        sum = 0.0
        for k = 1,10000 do
            sum = sum + 1.0/(k*k)
        end
    end
    return sum
end
]]
doast(str)

str=
[[local a: integer return a+3
]]
doast(str)

str=
[[local i: integer; return t[i/5]
]]
doast(str)

str=
[[local t: integer[]; return t[0][1]
]]
doast(str)

str=
[[return f()[1]
]]
doast(str)