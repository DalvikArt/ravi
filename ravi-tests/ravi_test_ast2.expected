function()
  return
    function()
    --locals  a, y, a, z
      local
      --[symbols]
        a --local symbol ? 
      --[expressions]
        1
      local
      --[symbols]
        y --local symbol function 
      --[expressions]
        function()
        --upvalues  a
          return
            function()
            --upvalues  a
              return
                --[primary start]
                  a --upvalue ? 
                --[primary end]
            end
        end
      local
      --[symbols]
        a --local symbol ? 
      --[expressions]
        5
      local
      --[symbols]
        z --local symbol function 
      --[expressions]
        function()
        --upvalues  a
          return
            function()
            --upvalues  a
              return
                --[primary start]
                  a --upvalue ? 
                --[primary end]
            end
        end
      return
        --[primary start]
          y --local symbol function 
        --[primary end]
       ,
        --[primary start]
          z --local symbol function 
        --[primary end]
    end
end

function()
  return
    function(
      a --local symbol ? 
     ,
      b --local symbol ? 
     ,
      c --local symbol ? 
     ,
      d --local symbol ? 
     ,
      e --local symbol ? 
    )
    --locals  a, b, c, d, e
      if
       --[binary expr start] any
        --[primary start]
          a --local symbol ? 
        --[primary end]
       ==
        --[primary start]
          b --local symbol ? 
        --[primary end]
       --[binary expr end]
      then
        goto l1
      elseif
       --[binary expr start] any
        --[primary start]
          a --local symbol ? 
        --[primary end]
       ==
        --[primary start]
          c --local symbol ? 
        --[primary end]
       --[binary expr end]
      then
        goto l2
      elseif
       --[binary expr start] any
        --[primary start]
          a --local symbol ? 
        --[primary end]
       ==
        --[primary start]
          d --local symbol ? 
        --[primary end]
       --[binary expr end]
      then
        goto l2
      else
        if
         --[binary expr start] any
          --[primary start]
            a --local symbol ? 
          --[primary end]
         ==
          --[primary start]
            e --local symbol ? 
          --[primary end]
         --[binary expr end]
        then
          goto l3
        else
          goto l3
        end
      end
      ::l1::
      ::l2::
      ::l3::
      ::l4::
    end
end

function()
  return
    function(
      a --local symbol ? 
    )
    --locals  a
      while
       --[binary expr start] any
        --[primary start]
          a --local symbol ? 
        --[primary end]
       <
        10
       --[binary expr end]
      do
        --[expression statement start]
         --[var list start]
           --[primary start]
             a --local symbol ? 
           --[primary end]
         = --[var list end]
         --[expression list start]
           --[binary expr start] any
            --[primary start]
              a --local symbol ? 
            --[primary end]
           +
            1
           --[binary expr end]
         --[expression list end]
        --[expression statement end]
      end
    end
end

function()
  return
    function(
      a --local symbol ? 
    )
    --locals  a
      ::L2::
      if
       --[unary expr start] any
       not
        --[primary start]
         --[binary expr start] any
          --[primary start]
            a --local symbol ? 
          --[primary end]
         <
          10
         --[binary expr end]
        --[primary end]
       --[unary expr end]
      then
        goto L1
      end
      --[expression statement start]
       --[var list start]
         --[primary start]
           a --local symbol ? 
         --[primary end]
       = --[var list end]
       --[expression list start]
         --[binary expr start] any
          --[primary start]
            a --local symbol ? 
          --[primary end]
         +
          1
         --[binary expr end]
       --[expression list end]
      --[expression statement end]
      goto L2
      ::L1::
    end
end

function()
  return
    function()
      --[expression statement start]
       --[expression list start]
         --[primary start]
          function()
          end
         --[primary end]
         --[suffix list start]
           --[function call start]
            (
              { --[table constructor start] table
                --[indexed assign start]
                --[value start]
                 --[primary start]
                   f --global symbol ? 
                 --[primary end]
                 --[suffix list start]
                   --[function call start]
                    (
                    )
                   --[function call end]
                 --[suffix list end]
                --[value end]
                --[indexed assign end]
              } --[table constructor end]
            )
           --[function call end]
         --[suffix list end]
       --[expression list end]
      --[expression statement end]
    end
end

function()
  return
    function()
    --locals  sum, j, k
      local
      --[symbols]
        sum --local symbol ? 
      for
        j --local symbol ? 
      =
        1
       ,
        500
      do
         --[expression statement start]
          --[var list start]
            --[primary start]
              sum --local symbol ? 
            --[primary end]
          = --[var list end]
          --[expression list start]
            0.0000000000000000
          --[expression list end]
         --[expression statement end]
         for
           k --local symbol ? 
         =
           1
          ,
           10000
         do
            --[expression statement start]
             --[var list start]
               --[primary start]
                 sum --local symbol ? 
               --[primary end]
             = --[var list end]
             --[expression list start]
               --[binary expr start] any
                --[primary start]
                  sum --local symbol ? 
                --[primary end]
               +
                --[binary expr start] any
                 1.0000000000000000
                /
                 --[primary start]
                  --[binary expr start] any
                   --[primary start]
                     k --local symbol ? 
                   --[primary end]
                  *
                   --[primary start]
                     k --local symbol ? 
                   --[primary end]
                  --[binary expr end]
                 --[primary end]
                --[binary expr end]
               --[binary expr end]
             --[expression list end]
            --[expression statement end]
         end
      end
      return
        --[primary start]
          sum --local symbol ? 
        --[primary end]
    end
end

